### PR TITLE
Removing `error` extension function

### PR DESCRIPTION
**Pull request by aaronjeline**
## Description of changes
Reduces the precision of partial evaluation when reducing expressions that are lazily evaluated (`if` `and` `or` subexpressions). The primary motivation for this is removing the `error` extension function, which is impossible to type.

Increases in precision without re-introducing `error` could come later.

## Issue #, if available
N/A
## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

File: cedar-policy/CHANGELOG.md [hunk 1]
```diff
@@ -25,6 +25,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Removed
 
+- Reduced precision of partial evaluation when evaluating lazy
+  expressions
```
Comment (khieta) [inline cedar-policy/CHANGELOG.md:29]: Probably makes more sense under the "changed" header than "removed" header. It also may be unclear to users what lazy expressions means in this context, so I'd suggest just
```suggestion
- Reduced precision of partial evaluation for `||`, `&&`,  and conditional expressions.
```

**Review by khieta — APPROVED** [review]
Looks fine to me. A little sad to see the precision loss in the updated test cases -- maybe open an issue to consider precision optimizations in the future?

File: cedar-policy/CHANGELOG.md [hunk 2]
```diff
@@ -25,6 +25,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Removed the `error` extension function
```
Comment (khieta) [inline cedar-policy/CHANGELOG.md:30]:
```suggestion
- Removed the `error` extension function, which was previously used during partial evaluation.
```

File: cedar-policy-core/src/ast/expr.rs [hunk 3]
```diff
@@ -321,6 +321,13 @@ impl Expr {
         ExprBuilder::new().ite(test_expr, then_expr, else_expr)
     }
 
+    /// Create a ternary (if-then-else) `Expr`.
+    /// Takes `Arc`s instead of owned `Expr`s
+    /// `test_expr` must evaluate to a Bool type
+    pub fn ite_arc(test_expr: Arc<Expr>, then_expr: Arc<Expr>, else_expr: Arc<Expr>) -> Self {
```
Comment (shaobo-he-aws) [inline cedar-policy-core/src/ast/expr.rs:327]: Nit: can we make it `pub(crate)`?

File: cedar-policy/CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Removed
 
+- Reduced precision of partial evaluation for `||`, `&&`,  and conditional expressions.
```
Comment (shaobo-he-aws) [inline cedar-policy/CHANGELOG.md:28]: Maybe we want to be more specific here? An example would help.

File: cedar-policy-core/src/ast/expr.rs (hunk 3 above)
Comment (aaronjeline) [reply cedar-policy-core/src/ast/expr.rs:327]: Rest of the functions are not `pub(crate)`, why this one?

Comment (shaobo-he-aws) [reply cedar-policy-core/src/ast/expr.rs:327]: I think the signature of this function is not idiomatic as public functions. I was thinking another function as well but just commented on this one.

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy-core/src/ast/expr.rs [hunk 5]
```diff
@@ -321,6 +321,13 @@ impl Expr {
         ExprBuilder::new().ite(test_expr, then_expr, else_expr)
     }
 
+    /// Create a ternary (if-then-else) `Expr`.
+    /// Takes `Arc`s instead of owned `Expr`s
```
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/expr.rs:325]:
```suggestion
    /// Takes `Arc`s instead of owned `Expr`s.
```

**Review by cdisselkoen — APPROVED** [review]

**Comment by aaronjeline** [comment]
Issue for re-increasing precision: https://github.com/cedar-policy/cedar/issues/880

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**